### PR TITLE
feat(site): add auto mode on create workspace form

### LINKED ISSUE #7717: open on coder button without subbmitting the form
is it possible to add and endpoint for create worksapce from a template by passing querry so that the user won't have to submit the form also would be nice to check if there is any workapce with the same name if there's it will open the vscode or whatever the template configure this way we wont have to wait for an integration like gitpods

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Closes #7717

### CODE REVIEW

**Comment by BrunoQuaresma** [comment]
Close https://github.com/coder/coder/issues/7717

**Comment by MrPeacockNLB** [comment]
Super cool! Can't await this feature!

Hopefully I'll find some time to public release the coder button as well. In conjuntion with this feature make really sense

**Comment by bpmct** [comment]
@matifali can take a look tomorrow :)

**Comment by matifali** [comment]
[Open in Coder](https://ss87nra5e58ae.pit-1.try.coder.app/templates/kubernetes/workspace?)

**Comment by matifali** [comment]
This gets me to this form
image

**Review by matifali — COMMENTED** [review]
@bpmct and @BrunoQuaresma I did deploy this, but I don't see any difference in the behavior compared to dev.coder.com
Maybe I am missing something. Do you know if I need to specify or enable something specifically?

**Review by matifali — DISMISSED** [review]
This gets me to this form
image

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 1]
```diff
@@ -1,27 +1,39 @@
 import { useMachine } from "@xstate/react"
```
Comment (mtojek) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: I did some basic testing of the feature, and it looks like "Open in Coder" link contains parameters that the user has clicked in the form.

Shouldn't it contain all values even if they are default?

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 2]
```diff
@@ -1,27 +1,39 @@
...
-import { TemplateVersionParameter } from "api/typesGenerated"
```
Comment (mtojek) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:2]: I'm not sure why but in my case "Open in Code" component is pretty big. Larger than the parameter form on the left. Is it intended?

Screenshot 2023-07-26 at 10 47 21

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 1 above)
Comment (mtojek) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: When the link contains an invalid parameter value, nothing is selected in the form. Should the parameter switch to the default value? Unchecked checkboxes look suspicious...

Screenshot 2023-07-26 at 10 49 10

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 3]
```diff
@@ -1,27 +1,39 @@
...
+import {
```
Comment (mtojek) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:2]: You may want to test it also with the "list of strings" type.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 4]
```diff
@@ -1,111 +1,103 @@
 import { useMachine } from "@xstate/react"
```
Comment (mtojek) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: The `mode=auto` seems to be work except for required parameters (without `default` property):

Screenshot 2023-07-26 at 11 14 09

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: I don't think so. If the user does not provide a value, I would expect we automatically get the default if it is provided. cc.: @bpmct wdyt?

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 2 above)
Comment (BrunoQuaresma) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:2]: Yes, it is. But I think I can improve that 🤔

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: Good catch, I agree with you.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 4 above)
Comment (BrunoQuaresma) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: This is a good one, thx for catching it.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx (hunk 1 above)
Comment (mtojek) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: What if default values change over time :) Do we want to stick to the one present at the link generation or use the fresh one?

Comment (bpmct) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:1]: If not specified, we should use the default of the template at the time the button was clicked, not generated. So if the default changes over time, different users will see different behaviors, but that is OK.

**Comment by bpmct** [comment]
Is there a way to specify a name via a URL parameter? If taken, it appends a random name?

**Review by mtojek — APPROVED** [review]
Looks good 👍 but there is one question from Ben that might be relevant.

**Comment by BrunoQuaresma** [comment]
@bpmct Yes, you can use the `?name=something` and if the name is taken you will be asked to add a new name but I don't see why this would be a good thing to do since the idea of a link would have multiple ppl using it, so the name only would work at the first time. 

Mergin it for now but I can do a hotfix if it is required. Ps: It would one more step to validate the name.

**Pull request merged**